src/ros_filter.cpp

```diff
@@ -2747,6 +2798,87 @@ namespace RobotLocalization
 
     return canTransform;
   }
+
+  template<typename T>
+  void RosFilter<T>::saveFilterState(FilterBase& filter)
+  {
+    FilterState state;
+    state.state_ = filter.getState();
+    state.estimate_error_covariance_ = filter.getEstimateErrorCovariance();
+    state.time_ = filter.getLastMeasurementTime();
+    filterStateHistory_.push_back(state);
+    ros::Time ts;
+    ts.fromSec(state.time_);
+    RF_DEBUG("\n---- RosFilter::saveFilterState for timestamp " << ts << " size = " << filterStateHistory_.size() << "----\n");
+  }
+
+  template<typename T>
+  bool RosFilter<T>::restoreFilterAndMeasurementsBefore(const double t)
+  {
+    RF_DEBUG("\n----- RosFilter::restoreFilterAndMeasurementsBefore (" << t << ") -----\n");
+    // Remove all filter states older than t. This yields either the latest
+    // filter state before t or an empty queue. In that case the smoothing
+    // window is too small and we cannot reset the filter.
+    //
+
+    int num_popped_filter_updates = 0;
+    while (!filterStateHistory_.empty()
+        && filterStateHistory_.back().time_ > t)
+    {
+      filterStateHistory_.pop_back();
```

Comment:
Correct me if I'm wrong, but it looks like if we get an old measurement, we destroy the state history. What if we get a measurement from sensor 1 that makes us go back 3 seconds, then we apply this logic and the filter gets updated correctly, but then we get a measurement that makes us go back 2 seconds? Won't we be in trouble? If so, would it make more sense to just iterate over the history and choose the first one that is before the time in question, without popping them?